corelibs/libcache.md

```diff
@@ -0,0 +1,187 @@
+```
+| Status | Description | Return value | Remarks |
+|--------|-------------|--------------| ------- |
+| Declared | Read callback — a pointer to a function responsible for reading data from the source memory. The pointer is registered in the cache during a call to `cache_init` function. <br /><br />The function reads up to `count` bytes from a source address `addr` into a `buffer`. <br /><br /> Utilizes additional device driver context provided in `ctx`.| **On success:** a number of bytes read from the cached source memory <br /><br /> **On failure:** an error number | **A cached source memory-specific implementation ought to be supplied by the user.** |
+
+```c
+typedef ssize_t (*cache_writeCb_t)(uint64_t offset, const void *buffer, size_t count, struct cache_devCtx_s *ctx);
+```
+| Status | Description | Return value | Remarks |
+|--------|-------------|--------------| ------- |
+| Declared | Write callback — a pointer to a function responsible for writing data to the source memory. The pointer is registered in the cache during a call to `cache_init` function. <br /><br /> The function writes up to `count` bytes from `buffer` to a source address `addr`. <br /><br /> Utilizes additional device driver context provided in `ctx`.| **On success:** a number of bytes written to the cached source memory<br/><br/>**On failure:** an error number <!-- TODO: write whether errno is set --> | **A cached source memory-specific implementation ought to be supplied by the user.** |
+
+```c
+cachectx_t *cache_init(size_t srcMemSize, size_t size, size_t lineSize, const cache_ops_t *ops);
+```
+| Status | Description | Return value | Remarks |
+|--------|-------------|--------------| ------- |
+|Implemented <br /><br />Tests in progress| Initializes the cache for use. Stores the size of the cached disk/flash as `srcMemSize` (in bytes). Sets the cache memory size to `size` and the cache line size to `lineSize` (both in bytes). <br /><br /> Registers cached source memory interface (`ops`) which contains write and read callbacks as well as device driver context. | **On success:** a pointer to initialized cache of type `cachectx_t` <br /><br /> **On failure:** NULL | `size` has to be a multiple of `lineSize`. <br /><br /> The `size` to `lineSize` ratio has to be a multiple of the number of ways in a set (see: [Associativity](#associativity)).
+
+```c
+int cache_deinit(cachectx_t *cache);
+```
+| Status | Description | Return value | Remarks |
+|--------|-------------|--------------| ------- |
+| Implemented <br /><br />Tests in progress | Writes dirty lines (see: [Write policy](#write-policy)) marked with dirty bit to the cached source memory and then deinitializes the cache: frees all allocated resources (cache lines, sets as well as the cache table). | **On success:** 0 <br /><br /> **On failure:** an error number | <!-- TODO: write about additional error codes, e.g. EBUSY --> May fail due to problems with the cached source memory (``EIO``).
```

Comment:
Performs cache flush and invalidation is enough, this is well established nomenclature when talking about cache